torchok/models/heads/detection/detr.py

```diff
@@ -0,0 +1,377 @@
+
+        if train_cfg:
+            if 'assigner' not in train_cfg:
+                raise ValueError("assigner should be provided when train_cfg is set.")
+            assigner = train_cfg['assigner']
+            self.assigner = build_assigner(assigner)
+            # DETR sampling=False, so use PseudoSampler
+            self.sampler = build_sampler(dict(type='PseudoSampler'), context=self)
+
+        self.num_query = num_query
+        self.num_classes = num_classes
+        self.in_channels = in_channels if isinstance(in_channels, int) else in_channels[0]
+        self.num_reg_fcs = num_reg_fcs
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+
+        self.act_cfg = transformer.get('act_cfg', dict(type='ReLU', inplace=True))
+        self.activate = build_activation_layer(self.act_cfg)
+        self.positional_encoding = build_positional_encoding(positional_encoding)
+        self.transformer = build_transformer(transformer)
+        self.embed_dims = self.transformer.embed_dims
+        num_feats = positional_encoding['num_feats']
+        if num_feats * 2 != self.embed_dims:
+            raise ValueError(
+                f'embed_dims should be exactly 2 times of num_feats. Found {self.embed_dims} and {num_feats}.')
+        self._init_layers()
+
+        self.init_weights()
+
+    def forward(self, features: List[torch.Tensor], img_metas: List[Dict] = None):
+        """Forward function.
+
+        Args:
+            features: Features from the upstream network, each is a 4D-tensor.
+            img_metas: List of image information.
+
+        Returns:
+            all_cls_scores (torch.Tensor): Outputs from the classification head,
+                shape [nb_dec, bs, num_query, cls_out_channels]. Note cls_out_channels should include background.
+            all_bbox_preds (torch.Tensor): Sigmoid outputs from the regression
+                head with normalized coordinate format (cx, cy, w, h). Shape [nb_dec, bs, num_query, 4].
+        """
+        features = features[-1]
+
+        x = self.input_proj(features)
+
+        batch_size = features.size(0)
+        if img_metas is not None:
+            input_img_h, input_img_w, _ = img_metas[0]['img_shape']
+            masks = features.new_ones((batch_size, input_img_h, input_img_w))
+            for img_id in range(batch_size):
+                img_meta = img_metas[img_id]
+                img_h, img_w = self.get_fit_image_size(img_meta['img_shape'], img_meta['orig_img_shape'])
+                dh = input_img_h - img_h
+                dw = input_img_w - img_w
+
+                masks[img_id, dh // 2:input_img_h + (-dh) // 2, dw // 2: input_img_w + (-dw) // 2] = 0
+            # interpolate masks to have the same spatial shape with x
+            masks = F.interpolate(masks.unsqueeze(1), size=x.shape[-2:]).to(torch.bool).squeeze(1)
+        else:
+            masks = torch.zeros((batch_size, *x.shape[-2:]), dtype=torch.bool, device=x.device)
```

Comment:
Why it should be zeros? Don't see this code in mmdetection

Reply:
This mask contains an inverted area of interest. If you provide image meta, we can obtain padded image borders and consider them as an area out of our interest, marking them as True or 1. Otherwise, if we don't get meta, we have to consider the whole image as an area of interest marking them as False or 0